// @terra-dev/use-wallet/useConnectedWallet.js
import { AccAddress } from '@terra-money/terra.js';
import { useMemo } from 'react';
import { ConnectType, WalletStatus } from './types';
import { useWallet } from './useWallet';
export function useConnectedWallet() {
    const { status, network, wallets, post, sign } = useWallet();
    const value = useMemo(() => {
        try {
            if (status === WalletStatus.WALLET_CONNECTED &&
                wallets.length > 0 &&
                AccAddress.validate(wallets[0].terraAddress)) {
                const { terraAddress, connectType, design } = wallets[0];
                return {
                    network,
                    terraAddress: terraAddress,
                    walletAddress: terraAddress,
                    design,
                    post: (tx) => {
                        return post(tx, { terraAddress });
                    },
                    sign: (tx) => {
                        return sign(tx, { terraAddress });
                    },
                    availablePost: connectType === ConnectType.WEB_CONNECT ||
                        connectType === ConnectType.CHROME_EXTENSION ||
                        connectType === ConnectType.WALLETCONNECT,
                    availableSign: connectType === ConnectType.CHROME_EXTENSION,
                    connectType,
                };
            }
            else {
                return undefined;
            }
        }
        catch (_a) {
            return undefined;
        }
    }, [network, post, sign, status, wallets]);
    return value;
}
