// @terra-money/wallet-provider/_commonjs/react/ExtensionNetworkOnlyWalletProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionNetworkOnlyWalletProvider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const use_wallet_1 = require("@terra-dev/use-wallet");
const react_1 = require("react");
const chrome_extension_1 = require("../modules/chrome-extension");
function ExtensionNetworkOnlyWalletProvider({ children, defaultNetwork, }) {
    const [controller] = (0, react_1.useState)(() => new chrome_extension_1.ChromeExtensionController({
        defaultNetwork,
        enableWalletConnection: false,
        dangerously__chromeExtensionCompatibleBrowserCheck: () => false,
    }));
    const [network, setNetwork] = (0, react_1.useState)(defaultNetwork);
    (0, react_1.useEffect)(() => {
        const networkSubscription = controller.networkInfo().subscribe({
            next: (value) => {
                setNetwork(value);
            },
        });
        return () => {
            networkSubscription.unsubscribe();
        };
    }, [controller]);
    const state = (0, react_1.useMemo)(() => {
        return {
            availableConnectTypes: [],
            availableInstallTypes: [],
            availableConnections: [],
            status: use_wallet_1.WalletStatus.WALLET_NOT_CONNECTED,
            network,
            wallets: [],
            install: () => {
                throw new Error('not implemented!');
            },
            connect: () => {
                throw new Error('not implemented!');
            },
            connectReadonly: () => {
                throw new Error('not implemented!');
            },
            disconnect: () => {
                throw new Error('not implemented!');
            },
            post: () => {
                throw new Error(`<ExtensionNetworkOnlyWalletProvider> does not support post()`);
            },
            sign: () => {
                throw new Error(`<ExtensionNetworkOnlyWalletProvider> does not support sign()`);
            },
            recheckStatus: controller.recheckStatus,
            isChromeExtensionCompatibleBrowser: () => {
                throw new Error('not implemented!');
            },
        };
    }, [controller.recheckStatus, network]);
    return ((0, jsx_runtime_1.jsx)(use_wallet_1.WalletContext.Provider, Object.assign({ value: state }, { children: children }), void 0));
}
exports.ExtensionNetworkOnlyWalletProvider = ExtensionNetworkOnlyWalletProvider;
