// @terra-money/wallet-provider/_commonjs/getChainOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getChainOptions = void 0;
const FALLBACK_MAINNET = {
    name: 'mainnet',
    chainID: 'columbus-5',
    lcd: 'https://lcd.terra.dev',
};
const FALLBACK = {
    defaultNetwork: FALLBACK_MAINNET,
    walletConnectChainIds: {
        1: FALLBACK_MAINNET,
        0: {
            name: 'testnet',
            chainID: 'bombay-12',
            lcd: 'https://bombay-lcd.terra.dev',
        },
    },
};
let cache;
async function getChainOptions() {
    return fetch('https://assets.terra.money/chains.json')
        .then((res) => res.json())
        .then((data) => {
        const chains = Object.values(data);
        const walletConnectChainIds = chains.reduce((result, network) => {
            if (typeof network.walletconnectID === 'number') {
                result[network.walletconnectID] = network;
            }
            else if (!result[1] && network.name === 'mainnet') {
                result[1] = network;
            }
            else if (!result[0] && network.name === 'testnet') {
                result[0] = network;
            }
            return result;
        }, {});
        const chainOptions = {
            defaultNetwork: walletConnectChainIds[1],
            walletConnectChainIds,
        };
        cache = chainOptions;
        return chainOptions;
    })
        .catch((error) => {
        console.error('Failed to fetch chains.json', error);
        return cache !== null && cache !== void 0 ? cache : FALLBACK;
    });
}
exports.getChainOptions = getChainOptions;
