// @terra-dev/use-wallet/_commonjs/useConnectedWallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useConnectedWallet = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const react_1 = require("react");
const types_1 = require("./types");
const useWallet_1 = require("./useWallet");
function useConnectedWallet() {
    const { status, network, wallets, post, sign } = (0, useWallet_1.useWallet)();
    const value = (0, react_1.useMemo)(() => {
        try {
            if (status === types_1.WalletStatus.WALLET_CONNECTED &&
                wallets.length > 0 &&
                terra_js_1.AccAddress.validate(wallets[0].terraAddress)) {
                const { terraAddress, connectType, design } = wallets[0];
                return {
                    network,
                    terraAddress: terraAddress,
                    walletAddress: terraAddress,
                    design,
                    post: (tx) => {
                        return post(tx, { terraAddress });
                    },
                    sign: (tx) => {
                        return sign(tx, { terraAddress });
                    },
                    availablePost: connectType === types_1.ConnectType.WEB_CONNECT ||
                        connectType === types_1.ConnectType.CHROME_EXTENSION ||
                        connectType === types_1.ConnectType.WALLETCONNECT,
                    availableSign: connectType === types_1.ConnectType.CHROME_EXTENSION,
                    connectType,
                };
            }
            else {
                return undefined;
            }
        }
        catch (_a) {
            return undefined;
        }
    }, [network, post, sign, status, wallets]);
    return value;
}
exports.useConnectedWallet = useConnectedWallet;
