// @terra-money/wallet-provider/react/ExtensionNetworkOnlyWalletProvider.js
import { jsx as _jsx } from "react/jsx-runtime";
import { WalletContext, WalletStatus } from '@terra-dev/use-wallet';
import { useEffect, useMemo, useState } from 'react';
import { ChromeExtensionController } from '../modules/chrome-extension';
export function ExtensionNetworkOnlyWalletProvider({ children, defaultNetwork, }) {
    const [controller] = useState(() => new ChromeExtensionController({
        defaultNetwork,
        enableWalletConnection: false,
        dangerously__chromeExtensionCompatibleBrowserCheck: () => false,
    }));
    const [network, setNetwork] = useState(defaultNetwork);
    useEffect(() => {
        const networkSubscription = controller.networkInfo().subscribe({
            next: (value) => {
                setNetwork(value);
            },
        });
        return () => {
            networkSubscription.unsubscribe();
        };
    }, [controller]);
    const state = useMemo(() => {
        return {
            availableConnectTypes: [],
            availableInstallTypes: [],
            availableConnections: [],
            status: WalletStatus.WALLET_NOT_CONNECTED,
            network,
            wallets: [],
            install: () => {
                throw new Error('not implemented!');
            },
            connect: () => {
                throw new Error('not implemented!');
            },
            connectReadonly: () => {
                throw new Error('not implemented!');
            },
            disconnect: () => {
                throw new Error('not implemented!');
            },
            post: () => {
                throw new Error(`<ExtensionNetworkOnlyWalletProvider> does not support post()`);
            },
            sign: () => {
                throw new Error(`<ExtensionNetworkOnlyWalletProvider> does not support sign()`);
            },
            recheckStatus: controller.recheckStatus,
            isChromeExtensionCompatibleBrowser: () => {
                throw new Error('not implemented!');
            },
        };
    }, [controller.recheckStatus, network]);
    return (_jsx(WalletContext.Provider, Object.assign({ value: state }, { children: children }), void 0));
}
