// @terra-dev/chrome-extension/storage.js
import { AccAddress } from '@terra-money/terra.js';
export const storage = typeof window === 'undefined' ? undefined : localStorage;
export const SESSION = '__terra_chrome_extension_session__';
export function getStoredSession() {
    const data = storage === null || storage === void 0 ? void 0 : storage.getItem(SESSION);
    if (!data) {
        return undefined;
    }
    try {
        const object = JSON.parse(data);
        if ('identifier' in object &&
            'walletAddress' in object &&
            AccAddress.validate(object['walletAddress'])) {
            return {
                identifier: object['identifier'],
                walletAddress: object['walletAddress'],
            };
        }
        else {
            storage === null || storage === void 0 ? void 0 : storage.removeItem(SESSION);
            return undefined;
        }
    }
    catch (_a) {
        storage === null || storage === void 0 ? void 0 : storage.removeItem(SESSION);
        return undefined;
    }
}
export function storeSession(session) {
    if (!AccAddress.validate(session.walletAddress)) {
        throw new Error(`${session.walletAddress} is invalid terra address!`);
    }
    storage === null || storage === void 0 ? void 0 : storage.setItem(SESSION, JSON.stringify(session));
}
export function clearSession() {
    storage === null || storage === void 0 ? void 0 : storage.removeItem(SESSION);
}
//export function getStoredAddress(): string | null {
//  const address = storage?.getItem(WALLET_ADDRESS);
//  return address && AccAddress.validate(address) ? address : null;
//}
//
//export function storeAddress(address: string) {
//  if (!AccAddress.validate(address)) {
//    throw new Error(`${address} is invalidate terra address!`);
//  }
//
//  storage && storage.setItem(WALLET_ADDRESS, address);
//}
//
//export function clearStore() {
//  storage && storage.removeItem(WALLET_ADDRESS);
//}
