// @terra-dev/chrome-extension/_commonjs/storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearSession = exports.storeSession = exports.getStoredSession = exports.SESSION = exports.storage = void 0;
const terra_js_1 = require("@terra-money/terra.js");
exports.storage = typeof window === 'undefined' ? undefined : localStorage;
exports.SESSION = '__terra_chrome_extension_session__';
function getStoredSession() {
    const data = exports.storage === null || exports.storage === void 0 ? void 0 : exports.storage.getItem(exports.SESSION);
    if (!data) {
        return undefined;
    }
    try {
        const object = JSON.parse(data);
        if ('identifier' in object &&
            'walletAddress' in object &&
            terra_js_1.AccAddress.validate(object['walletAddress'])) {
            return {
                identifier: object['identifier'],
                walletAddress: object['walletAddress'],
            };
        }
        else {
            exports.storage === null || exports.storage === void 0 ? void 0 : exports.storage.removeItem(exports.SESSION);
            return undefined;
        }
    }
    catch (_a) {
        exports.storage === null || exports.storage === void 0 ? void 0 : exports.storage.removeItem(exports.SESSION);
        return undefined;
    }
}
exports.getStoredSession = getStoredSession;
function storeSession(session) {
    if (!terra_js_1.AccAddress.validate(session.walletAddress)) {
        throw new Error(`${session.walletAddress} is invalid terra address!`);
    }
    exports.storage === null || exports.storage === void 0 ? void 0 : exports.storage.setItem(exports.SESSION, JSON.stringify(session));
}
exports.storeSession = storeSession;
function clearSession() {
    exports.storage === null || exports.storage === void 0 ? void 0 : exports.storage.removeItem(exports.SESSION);
}
exports.clearSession = clearSession;
//export function getStoredAddress(): string | null {
//  const address = storage?.getItem(WALLET_ADDRESS);
//  return address && AccAddress.validate(address) ? address : null;
//}
//
//export function storeAddress(address: string) {
//  if (!AccAddress.validate(address)) {
//    throw new Error(`${address} is invalidate terra address!`);
//  }
//
//  storage && storage.setItem(WALLET_ADDRESS, address);
//}
//
//export function clearStore() {
//  storage && storage.removeItem(WALLET_ADDRESS);
//}
