// @terra-dev/web-connector-interface/errors.js
export class WebConnectorUserDenied extends Error {
    constructor() {
        super('User Denied');
        this.toString = () => {
            return `[${this.name}]`;
        };
        this.toJSON = () => {
            return {
                name: this.name,
            };
        };
        this.name = 'WebConnectorUserDenied';
    }
}
export class WebConnectorCreateTxFailed extends Error {
    constructor(message) {
        super(message);
        this.toString = () => {
            return `[${this.name} message="${this.message}"]`;
        };
        this.toJSON = () => {
            return {
                name: this.name,
                message: this.message,
            };
        };
        this.name = 'WebConnectorCreateTxFailed';
    }
}
export class WebConnectorTxFailed extends Error {
    constructor(txhash, message, raw_message) {
        super(message);
        this.txhash = txhash;
        this.raw_message = raw_message;
        this.toString = () => {
            return `[${this.name} txhash="${this.txhash}" message="${this.message}"]\n${JSON.stringify(this.raw_message, null, 2)}`;
        };
        this.toJSON = () => {
            return {
                name: this.name,
                txhash: this.txhash,
                message: this.message,
                raw_message: this.raw_message,
            };
        };
        this.name = 'WebConnectorTxFailed';
    }
}
export class WebConnectorTxUnspecifiedError extends Error {
    constructor(message) {
        super(message);
        this.toString = () => {
            return `[${this.name} message="${this.message}"]`;
        };
        this.toJSON = () => {
            return {
                name: this.name,
                message: this.message,
            };
        };
        this.name = 'WebConnectorTxUnspecifiedError';
    }
}
export class WebConnectorLedgerError extends Error {
    constructor(code, message) {
        super(message);
        this.code = code;
        this.toString = () => {
            return `[${this.name} code="${this.code}" message="${this.message}"]`;
        };
        this.toJSON = () => {
            return {
                name: this.name,
                code: this.code,
                message: this.message,
            };
        };
        this.name = 'WebConnectorLedgerError';
    }
}
// ---------------------------------------------
// functions
// ---------------------------------------------
export function isWebConnectorError(error) {
    return (error instanceof WebConnectorUserDenied ||
        error instanceof WebConnectorCreateTxFailed ||
        error instanceof WebConnectorTxFailed ||
        error instanceof WebConnectorLedgerError ||
        error instanceof WebConnectorTxUnspecifiedError);
}
export function createTxErrorFromJson(json) {
    switch (json.name) {
        case 'WebConnectorUserDenied':
            return new WebConnectorUserDenied();
        case 'WebConnectorCreateTxFailed':
            return new WebConnectorCreateTxFailed(json.message);
        case 'WebConnectorLedgerError':
            return new WebConnectorLedgerError(json.code, json.message);
        case 'WebConnectorTxFailed':
            return new WebConnectorTxFailed(json.txhash, json.message, json.raw_message);
        default:
            return new WebConnectorTxUnspecifiedError('message' in json ? json.message : String(json));
    }
}
